WIP / TST: allow testing torch._numpy under Dynamo

ev-br opened the pull request:
Use conditional imports: when running under dynamo, import the original NumPy not torch._numpy. This is what we want to trace, not our implementation.

With this, the test suite passes with and without `PYTORCH_TEST_WITH_DYNAMO=1` (modulo a couple of test modules which are not meant to be compiled, e.g. `test_nep50_examples`). There are two new decorators, `x{fail,pass}ifTorchDynamo`, the `xpass` in most cases indicates a graph break and a fallback to eager for things we do not implement. 

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ev-br committed:
TST: allow testing torch._numpy under Dynamo

Use conditional imports: when running under dynamo, import the original
NumPy not torch._numpy. This is what we want to trace, not our implementation.

lezcano commented on test/torch_np/numpy_tests/core/test_multiarray.py:
```diff
@@ -22,31 +22,52 @@
 )
 
+
+if TEST_WITH_TORCHDYNAMO:
+    import numpy as np
+    from numpy.testing import (
+        assert_,
+        assert_allclose,  # IS_PYPY, IS_PYSTON, HAS_REFCOUNT,
+        assert_almost_equal,
+        assert_array_almost_equal,
+        assert_array_equal,
+        assert_array_less,
+        assert_equal,
+        assert_raises_regex,
+        assert_warns,
+        # runstring, temppath,
+        suppress_warnings,  # break_cycles,
+    )
+
+else:
+    import torch._numpy as np
+    from torch._numpy.testing import (
+        assert_,
+        assert_allclose,  # IS_PYPY, IS_PYSTON, HAS_REFCOUNT,
+        assert_almost_equal,
+        assert_array_almost_equal,
+        assert_array_equal,
+        assert_array_less,
+        assert_equal,
+        assert_raises_regex,
+        assert_warns,
+        # runstring, temppath,
+        suppress_warnings,  # break_cycles,
+    )
```
Could we simply do:
```suggestion
# If we are going to trace through these, we should use NumPy
# If testing on eager mode, we use torch._numpy
if TEST_WITH_TORCHDYNAMO:
    import numpy as np
else:
    import torch._numpy as np

from np.testing import (
    assert_,
    assert_allclose,  # IS_PYPY, IS_PYSTON, HAS_REFCOUNT,
    assert_almost_equal,
    assert_array_almost_equal,
    assert_array_equal,
    assert_array_less,
    assert_equal,
    assert_raises_regex,
    assert_warns,
    # runstring, temppath,
    suppress_warnings,  # break_cycles,
)
```

lezcano replied:
That's annoying.

lezcano replied:
Leaving a comment similar to the one I wrote would be nice, tho.

lezcano commented on test/torch_np/numpy_tests/core/test_multiarray.py:
```diff
@@ -1804,9 +1825,9 @@ def test_argsort_axis(self):
-            np.array([0, 1, np.nan], dtype=np.float16),
-            np.array([0, 1, np.nan], dtype=np.float32),
-            np.array([0, 1, np.nan]),
+            subtest(np.array([0, 1, np.nan], dtype=np.float16), name='f16'),
+            subtest(np.array([0, 1, np.nan], dtype=np.float32), name='f32'),
+            subtest(np.array([0, 1, np.nan]), name='default_dtype')
```
Why do we keep finding instances of these?

ev-br replied:
We did not run with `np` being numpy before, and names/reprs are slightly different (apparently).

ev-br committed:
TST: skip a test under Dynamo (complex inf/nans differ w.r.t. eager)

Here the test compares pytorch.logaddexp to scipy.special.logsumexp,
dynamo traces the latter and hits the inf/nan difference. So under dynamo
the reference values are wrong, not the ones being tested.

ev-br committed:
TST: graph break in torch._numpy.convolve

FIxes under Dynamo

python test/nn/test_convolution.py -k test_conv1d_vs_scipy_mode_same_cpu_complex6

ev-br committed:
TST: skip a Caffe2 tensorboard test with float128 under Dynamo

ev-br committed:
TST: make tests in numpy_tests/core dynamo-ready

ev-br committed:
TST: allow running torch_np FFT tests under Dynamo

ev-br committed:
TST: allow running torch_np linalg tests under Dynamo

While at it,

- stop skipping slow tests
- fix testing to zero-sized arrays (these seem to work, actually)

ev-br committed:
TST: allow tracing through lib/test_arraypad

ev-br committed:
TST: allow tracing through lib/test_arrraysetops

ev-br committed:
TST: lib/test_function_base

ev-br committed:
TST: lib/test_index_tricks

ev-br committed:
TST: lib/test_shape_base

ev-br committed:
lib: twodimbase

ev-br committed:
lib/test_type_check

ev-br committed:
oh mighty linter

ev-br committed:
lint

ev-br committed:
TST: trivially refactor test_twodim_base to avoid confusing Dynamo guards

Otherwise, it emits `RuntimeError: Could not infer dtype of NoneType`.
Admittedly, the construction was weird, so not entirely surprising that it confuses
the compiler.

ev-br committed:
TST: skip a test under dynamo (array scalars vs 0D arrays, again)

ev-br committed:
lint

ev-br committed:
add {xfail,xpass}IfTorchDynamo

ev-br committed:
TST: make numpy_tests/lib tests mostly pass under dynamo

ev-br committed:
more xpassIfTorchDynamo

ev-br committed:
TST: fix minor test_scalarmath under dynamo

ev-br committed:
TST: mark all xpasses in numpy_tests/core

ev-br committed:
TST: mark xpassIfTorchDynamo in numpy_tests/linalg

ev-br committed:
TST: xpassIfTorchDynamo numpy_tests/lib/

ev-br committed:
TST: xpassIfTorchDynamo tests/

ev-br committed:
also reductions

ev-br committed:
lint

ev-br committed:
CI: stop downloading/installing numpy_torch_interop

ev-br committed:
TST: remove a not useful test (fails CI for no good reason)

ev-br committed:
TST: skip a test on older numpies

Older numpies emit a DeprecationWarning. Not sure about the exact version, so put
in what is known to work.

ev-br committed:
TST: restrict the dtypes (spurious CI failure)

lezcano commented on test/torch_np/numpy_tests/fft/test_pocketfft.py:
```diff
@@ -358,10 +365,13 @@ def worker(args, q):
         # Make sure all threads returned the correct value
         for i in range(self.threads):
-            assert_array_equal(
+            # under torch.dynamo `assert_array_equal` fails with relative errors of
+            # about 1.5e-14. Hence replace it with `assert_allclose(..., rtol=2e-14)`
```
thank you for all these comments!

lezcano commented on test/torch_np/numpy_tests/lib/test_arraysetops.py:
```diff
@@ -217,7 +231,7 @@ def _isin_slow(a, b):
 
         def assert_isin_equal(a, b):
-            x = isin(a, b, kind=kind)
+            x = np.isin(a, b, kind=kind)
```
what happened here before?

ev-br replied:
The original numpy tests had `from numpy import isin`; when porting this we skipped/xfailed all tests which required it, and it never ran. We just never noticed that the xfail reason was an ImportError (or it was masked by other ImportErrors).

Now with a conditional import, we either need to add `If TEST_WITH_TORCHDYNAMO: from numpy import isin`, or do this line.

So in this PR, this will graph break (because we do not implement `torch._numpy.isin`).

lezcano replied:
thank you for the explanation!

lezcano commented on test/torch_np/numpy_tests/lib/test_function_base.py:
```diff
@@ -2537,7 +2597,19 @@ def test_error_not_1d(self, vals):
+)
+
+
+@xpassIfTorchDynamo  # (reason="TODO: implement")
+@instantiate_parametrized_tests
```
how come the script missed these? Could we fix the script to detect these?

ev-br replied:
This was hiding in the false negatives output of the script, I guess.

lezcano commented on test/torch_np/test_reductions.py:
```diff
@@ -302,6 +316,10 @@ def test_sum_where(self):
     np.count_nonzero,
 }
 
+restricts_dtype_casts = {np.var, np.std}
+
+fails_empty_tuple = {np.argmin, np.argmax}
```
These we can fix later right?

ev-br replied:
`fails...` is not the right word probably. We allow it while numpy does not:

```
In [1]: import numpy as np

In [2]: np.argmin(np.ones(3), axis=())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/numpy/core/fromnumeric.py:59, in _wrapfunc(obj, method, *args, **kwds)
     58 try:
---> 59     return bound(*args, **kwds)
     60 except TypeError:
     61     # A TypeError occurs if the object does have such a method in its
     62     # class, but its signature is not identical to that of NumPy's. This
   (...)
     66     # Call _wrapit from within the except clause to ensure a potential
     67     # exception has a traceback chain.

TypeError: 'tuple' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 np.argmin(np.ones(3), axis=())

<snip>

File ~/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/numpy/core/fromnumeric.py:45, in _wrapit(obj, method, *args, **kwds)
     43 except AttributeError:
     44     wrap = None
---> 45 result = getattr(asarray(obj), method)(*args, **kwds)
     46 if wrap:
     47     if not isinstance(result, mu.ndarray):

TypeError: 'tuple' object cannot be interpreted as an integer

```

lezcano reviewed (approved):
Mostly comments, once CI is green we can merge.

Thank you for this massive effort, this will give us a better signal when we change things internally in dynamo

lezcano commented on pytest.ini:
```diff
@@ -13,3 +13,5 @@ testpaths =
 junit_logging_reruns = all
 filterwarnings =
     ignore:Module already imported so cannot be rewritten.*hypothesis:pytest.PytestAssertRewriteWarning
+
+strict_xfail = True
```
I mentioned we can merge once ci is green, but actually I'd like to know if this line is strictly necessary.

ev-br replied:
TL;DR: not strictly necessary, is a technical implement, is safe to remove

With this, xpass (a test is marked as xfail but passes) is reported as a failure. Is useful if the test suite has ~50 unexpected xpasses---with this, they are all reported at the bottom of the log; otherwise, need to fish out them manually.

Also, this is only active if run via `pytest path/to/test/file` and not if run via pytorch-y `python path/to/test/file`.

lezcano replied:
SGTM, but tagging @ezyang for confirmation

ezyang replied:
`strict_xfail` isn't a thing. Did you mean `xfail_strict`?

ezyang replied:
The defaults change is fine. I actually thought pytest defaulted this way. Note that some devs do use pytest for local dev, it is a supported flow.

ev-br commented on test/torch_np/numpy_tests/lib/test_twodim_base.py:
```diff
@@ -127,9 +159,10 @@ def test_vector(self):
-        if vals is None:
-            vals = (100 * get_mat(5) + 1).astype("l")
+    def test_matrix(self):
+        self.check_matrix(vals=(100 * get_mat(5) + 1).astype("l"))
```
Dynamo gets confused by the  ndarray or none polymorphism.

lezcano replied:
Best to xfail these things rather than fix them by hand, but sure, we'll just land the fix soon

ev-br replied:
Two reasons:
- the pattern is rather weird, and only works by chance. Basically, it's `test_` function with an optional parameter, which is called twice: once as a regular function via test collection, and is called with an explicit parameter from inside some other `test_` function.
- both tests are actually good.
I'd much rather have a `check_` function called by two separate `test_` functions. Saving a single `def` LOC does not seem to be worth it IMO. As long as it worked, not worth tweaking on its own; but once I looked at this anyway, seems to be a good time to fix a wart and move on.

lezcano replied:
It's just that, in general, it's better to fix things on dynamo than manually, but sure.

ev-br commented on test/torch_np/numpy_tests/core/test_scalarmath.py:
```diff
@@ -317,7 +332,7 @@ def test_float_modulus_exact(self):
     def test_float_modulus_roundoff(self):
         # gh-6127
-        dt = np.typecodes["Float"]
+        dt = "efd"
```
Commented-out code or a patch in the GH history?

lezcano replied:
not sure what you mean by the latter? My point is that it'd be good to keep the original one once we fix it in dynamo, which should be quite easy with the pattern in your other PR

ev-br committed:
CI: rm commit pin for numpy_pytorch_interop

ev-br committed:
BUG: special-case np.__version__ (do not trace, it is always NumPy)

ev-br committed:
TST: add a reason to a skip

ev-br commented on .ci/pytorch/common_utils.sh:
```diff
@@ -171,13 +171,6 @@ function install_torchrec_and_fbgemm() {
 
-function install_numpy_pytorch_interop() {
-  local commit
-  commit=$(get_pinned_commit numpy_pytorch_interop)
```
done

ev-br committed:
numpy.__version__ not np.__version__ on eager

ev-br committed:
TST: add a missing instantiate_parametrized_tests

ev-br committed:
restore commented out `np.typecodes["AllInteger"]

ev-br committed:
TST: fix the last crop of CI failures

ev-br committed:
more CI fails

ev-br committed:
TST: more CI fails

ev-br committed:
does it ever end

lezcano commented:
FWIW, once CI is green (after a rebase), feel free to trigger the merge

ev-br committed:
does not apparently

ev-br committed:
TST: skip tests which fail on NumPy 1.23

ev-br committed:
more CI failures

ev-br committed:
clang10 fails on CI

ev-br committed:
more

lezcano commented on test/torch_np/numpy_tests/linalg/test_linalg.py:
```diff
@@ -802,7 +841,7 @@ def test_singular(self):
             linalg.cond(A, p)
 
-    @xfail  # (
+    @skip(reason="NPY_VER: fails on CI")  # (
```
some say NPY_VER, other NP_VER and others just fail on CI. Let's just use one of them so that we can consistently grep for all of them

ev-br committed:
consistent comments

ev-br commented:
@pytorchbot merge -r

ev-br committed:
one more skip

lezcano commented:
@pytorchbot merge -r

lezcano commented:
CI looks green, let's merge it quickly!

lezcano commented:
@pytorchbot merge -r

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but it is failing dynamo job in trunk https://hud.pytorch.org/pytorch/pytorch/commit/5ed4a423ded14138f1a724eff15ccd14648f6c49' -c weird

ev-br commented:
OK, I'm breaking this up into smaller PRs

ev-br commented:
closing in favor of the stack based off https://github.com/pytorch/pytorch/pull/112057
